Duplicate Binding Error Incorrectly Masked

``` java
public class Foo {
  public static void main(String... args) {
    ObjectGraph.create(new FooModuleOne(), new FooModuleTwo());
  }

  @Module
  static class FooModuleOne {
  }

  @Module(overrides = false, library = true, complete = false)
  static class FooModuleTwo {
    @Provides @Singleton @Test String provideA() {
      return "A";
    }

    @Provides @Singleton @Test String provideB() {
      return "B";
    }
  }

  @Qualifier
  @interface Test {
  }
}
```

```
Caused by: java.lang.IllegalArgumentException: FooModuleTwo is an overriding module and cannot contribute set bindings.
    at dagger.ObjectGraph$DaggerObjectGraph.makeGraph(ObjectGraph.java:187)
    at dagger.ObjectGraph$DaggerObjectGraph.access$000(ObjectGraph.java:138)
    at dagger.ObjectGraph.create(ObjectGraph.java:129)
    at coffee.Foo.main(Foo.java:11)
    ... 6 more
```

I might give this a go tonight. But no promises!

--- a/core/src/main/java/dagger/ObjectGraph.java
+++ b/core/src/main/java/dagger/ObjectGraph.java
@@ -184,8 +184,8 @@ public abstract class ObjectGraph {
           BindingsGroup addTo = moduleAdapter.overrides ? overrideBindings : baseBindings;
           moduleAdapter.getBindings(addTo, loadedModule.getValue());
         } catch (IllegalArgumentException e) {
-          throw new IllegalArgumentException(moduleAdapter.moduleClass.getSimpleName()
-              + " is an overriding module and cannot contribute set bindings.");
+          throw new IllegalArgumentException(
+              moduleAdapter.moduleClass.getSimpleName() + ": " + e.getMessage(), e);
         }
       }
 

--- a/core/src/test/java/dagger/ExtensionWithSetBindingsTest.java
+++ b/core/src/test/java/dagger/ExtensionWithSetBindingsTest.java
@@ -28,6 +28,7 @@ import org.junit.runners.JUnit4;
 import static dagger.Provides.Type.SET;
 import static org.fest.assertions.Assertions.assertThat;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
 @RunWith(JUnit4.class)
@@ -56,6 +57,16 @@ public final class ExtensionWithSetBindingsTest {
     @Provides(type=SET) @Singleton Integer provideD() { return counter.getAndIncrement(); }
   }
 
+  @Module
+  static class EmptyModule {
+  }
+
+  @Module(library = true)
+  static class DuplicateModule {
+    @Provides @Singleton String provideFoo() { return "foo"; }
+    @Provides @Singleton String provideBar() { return "bar"; }
+  }
+
   @Test public void basicInjectionWithExtension() {
     ObjectGraph root = ObjectGraph.createWith(new TestingLoader(), new RootModule());
     RealSingleton rs = root.get(RealSingleton.class);
@@ -81,8 +92,16 @@ public final class ExtensionWithSetBindingsTest {
       ObjectGraph.createWith(new TestingLoader(), new RootModule()).plus(new TestModule());
       fail("Should throw exception.");
     } catch (IllegalArgumentException e) {
-      assertEquals("TestModule is an overriding module and cannot contribute set bindings.",
-          e.getMessage());
+      assertEquals("TestModule: Module overrides cannot contribute set bindings.", e.getMessage());
+    }
+  }
+
+  @Test public void duplicateBindingsInSecondaryModule() {
+    try {
+      ObjectGraph.createWith(new TestingLoader(), new EmptyModule(), new DuplicateModule());
+      fail("Should throw exception.");
+    } catch (IllegalArgumentException e) {
+      assertTrue(e.getMessage().startsWith("DuplicateModule: Duplicate"));
     }
   }
 }

